File: 0124-binary-tree-maximum-path-sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.res = float("-inf")
        self.dfs(root)
        return self.res
    def dfs(self, root):
        if not root:
            return False
        
        leftmax = max(self.dfs(root.left), 0)
        rightmax = max(self.dfs(root.right), 0)
        self.res = max(self.res, leftmax + rightmax + root.val)
        
        return root.val + max(leftmax, rightmax)

File: 0124-binary-tree-maximum-path-sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res;
    int dfs(TreeNode* node){
        if (not node){
            return 0;
        }
        int leftmax = max(dfs(node->left), 0);
        int rightmax = max(dfs(node->right), 0);
        //path where we split
        res = max(res, leftmax + rightmax + node->val);
        //path where we do NOT split I.E. continue with largest path;
        return (node->val + max(leftmax, rightmax));

    }
    int maxPathSum(TreeNode* root) {
        res = root->val;
        dfs(root);
        return res; 
    }
};
